File: onestep00/828_Count_Unique_Characters_of_All_Substrings_of_a_Given_String.py
class Solution:
    def uniqueLetterString(self, s: str) -> int:
        import string
        
        # 각 문자 처음 위치
        index = {c: [-1, -1] for c in string.ascii_uppercase}
        res = 0
        
        # 각 위치의 문자가 언제 unique 할 수 있는지
        for i, c in enumerate(s):
            k, j = index[c]
            
            # 이전 위치에서 지금 위치까지 가능한 경우의 수
            res += (i - j) * (j - k)
            index[c] = [j, i]
            
        # 마지막까지
        for c in index.keys():
            k, j = index[c]
            res += (len(s) - j) * (j - k)
        return res

File: t5271688/828. Count Unique Characters of All Substrings of a Given String.cpp
class Solution {
public:
    int uniqueLetterString(string s) {
        vector<int> V[26];
        int ans=0;
        int n=s.length();
        for(int i=0; i<26; ++i) V[i].push_back(-1);
        for(int i=0; i<n; ++i) {
            V[s[i]-'A'].push_back(i);
        }
        for(int i=0; i<26; ++i) {
            V[i].push_back(n);
            for(int j=1; j<V[i].size()-1; j++) {
                ans+=(V[i][j]-V[i][j-1])*(V[i][j+1]-V[i][j]);
            }
        }
        return ans;
    }
};